### ISSUE TITLE
Hypergeometric series does not converge

### ISSUE THREAD

**Issue by barneyhill** [tskit-dev/tsdate#322]
Hmm just encountered this error - any ideas on how I could fix this issue?

```
(tsinfer) dnanexus@job-GZP5vx8Jg8JvG1Jj35f6763G:~/ukb/workflow-dask$ tsdate date UKB_EUR-chr21-filter_minimal-truncate-0.4-0.6-1-mm0-processed-simplified.trees UKB_EUR-chr21-filter_minimal-truncate-0.4-0.6-1-mm0-processed-simplified-dated.trees 10000 -m 1e-8 --method variational_gamma --progress

Find Node Spans: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 11959/11959 [2:43:23<00:00,  1.22it/s]
TipCount: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [2:43:23<00:00, 3267.99s/it]
2023-10-05 10:28:00,780 WARNING Initialising your tsdate installation by creating a user cache of conditional coalescent prior values for 1000 tips
Find Mixture Priors: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 669539/669539 [00:03<00:00, 182517.72it/s]

Expectation Propagation:   0%|                                                                                                                             | 0/20 [00:30<?, ?it/s]
Traceback (most recent call last):
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/approx.py", line 280, in gamma_projection
    logconst, t_i, ln_t_i, t_j, ln_t_j = sufficient_statistics(
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/hypergeo.py", line 200, in _hyp2f1_taylor_series
    raise Invalid2F1("Maximum terms reached in Taylor series")
tsdate.hypergeo.Invalid2F1: Maximum terms reached in Taylor series

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/approx.py", line 292, in gamma_projection
    logconst, t_i, va_t_i, t_j, va_t_j = mean_and_variance(
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/approx.py", line 246, in mean_and_variance
    v0 = mpmath.hyp2f1(a, b, c, z, maxterms=maxterms)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 251, in hyp2f1
    return ctx.hyper([a,b],[c],z,**kwargs)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 226, in hyper
    if   q == 1: return ctx._hyp2f1(a_s, b_s, z, **kwargs)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 456, in _hyp2f1
    v = ctx.hypercomb(h, [a,b], **kwargs)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 125, in hypercomb
    v = ctx.fprod([ctx.hyper(a_s, b_s, z, **kwargs)] + \
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 226, in hyper
    if   q == 1: return ctx._hyp2f1(a_s, b_s, z, **kwargs)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/functions/hypergeometric.py", line 443, in _hyp2f1
    return ctx.hypsum(2, 1, (atype, btype, ctype), [a, b, c], z, **kwargs)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/mpmath/ctx_mp.py", line 714, in hypsum
    zv, have_complex, magnitude = summator(coeffs, v, prec, wp, \
  File "<string>", line 48, in hypsum_2_1_RR_R_R
mpmath.libmp.libhyper.NoConvergence: Hypergeometric series converges too slowly. Try increasing maxterms.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dnanexus/ukb/tsinfer/bin/tsdate", line 8, in <module>
    sys.exit(main())
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/__main__.py", line 5, in main
    cli.tsdate_main()
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/cli.py", line 236, in tsdate_main
    args.runner(args)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/cli.py", line 217, in run_date
    dated_ts = tsdate.date(ts, args.mutation_rate, args.population_size, **params)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/core.py", line 1309, in date
    tree_sequence, dates, posteriors, timepoints, eps, nds, lik = variational_dates(
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/core.py", line 1631, in variational_dates
    dynamic_prog.iterate(iter_num=it, max_shape=max_shape)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/core.py", line 1115, in iterate
    self.propagate(
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/core.py", line 1067, in propagate
    ) = approx.gamma_projection(*parent_cavity, *child_cavity, *edge_lik)
  File "/home/dnanexus/ukb/tsinfer/lib/python3.9/site-packages/tsdate/approx.py", line 298, in gamma_projection
    raise hypergeo.Invalid2F1(
tsdate.hypergeo.Invalid2F1: Hypergeometric series does not converge; the approximate marginal is likely degenerate.  This may reflect a topological constraint that is at odds with the mutational data. Setting 'max_shape' to a large value (e.g. 1000) will prevent degenerate marginals, but the results should be treated with care.
```

Many Thanks,
Barney

**Comment by hyanwong** [comment]
Is this a fully simplified tree sequence, or does it have nodes with unary segments?

**Comment by barneyhill** [comment]
This is simplified with ts.simplify() prior to tsdate.

**Comment by hyanwong** [comment]
Ah, so it's not a unary segment issue. You can do what it says and set `max_shape`, but it might well be indicating a weirdness with the tree sequence you inferred, so it's worth looking into that too. Have you looked at the output of tsinfer with tsqc?

**Comment by nspope** [comment]
Setting `max_shape=100` should let it run through. However, the error indicates something strange might be going on with the topologies. To be clear on what it means, effectively we've hit an edge where the current estimate for parent age is very inconsistent with the current estimate for child age (e.g. it's much younger). This can happen if two parts of the tree sequence tell a very different story regarding the same node. 

But if it's just a few updates in the initial iteration, that's nothing to worry about (estimates typically take 1-2 iterations to stabilize). I would run with
```
dated_ts, posteriors = tsdate.date(ts, ..., max_shape=1000, return_posteriors=True)
```
"posteriors" will give you the raw gamma posteriors (not point estimates that have been adjusted to satisfy branch length constraints). Then,
```
node_ages_raw = np.zeros(ts.num_nodes)
node_ages_shape = np.zeros(ts.num_nodes)
for n in range(ts.num_samples, ts.num_nodes):
    node_ages_raw[n] = posteriors[n][0] / posteriors[n][1]
    node_ages_shape[n] = posteriors[n][0]
```
then it's instructive to plot these against the level order (the node times that are output by tsinfer, prior to dating). The relationship should look monotonic (like the second figure in this [comment](https://github.com/tskit-dev/tsdate/issues/288#issuecomment-1650444589), and *not* like the first).

**Comment by barneyhill** [comment]
Me and Jerome had a look at my chr21 exome+array trees with tsqc and apparently they didn't look great at all... I'm considering switching over to running tsinfer+tsdate on WGS data now.

**Comment by hyanwong** [comment]
We could do with some screen grabs for https://github.com/tskit-dev/tsqc/issues/98 (assuming that that doesn't break UKB rules)

**Comment by nspope** [comment]
This is no longer relevant after #345, we can close
